src/middleware.ts

```diff
@@ -1,106 +1,124 @@
-import { AzureFunction, Context } from '@azure/functions';
+import { FunctionHandler, InvocationContext } from '@azure/functions';
 
 import { errorHandler } from './error';
+import { stringify } from './util/stringify';
 
-type ErrorResult = { $failed: true; $error: unknown };
+export type ExceptionalResult = { $failed: true; $error: Error };
+export type MiddlewareResult<T> = ExceptionalResult | { $failed: false; $result: Awaited<T | undefined> };
 
-const isErrorResult = (result: unknown | ErrorResult): result is ErrorResult => (result as ErrorResult)?.$failed;
+export type BeforeExecutionFunction<T = FunctionHandler> = T extends (...a: infer U) => infer R
+    ? (...a: [...U, MiddlewareResult<R>]) => unknown
+    : never;
+
+export type PostExecutionFunction<T = FunctionHandler> = BeforeExecutionFunction<T>;
+
+export const isErrorResult = <T>(result: MiddlewareResult<T> | ExceptionalResult): result is ExceptionalResult =>
+    (result as ExceptionalResult)?.$failed;
 
 const middlewareCore =
-    (beforeExecution: (AzureFunction | false)[], handler: AzureFunction, postExecution: (AzureFunction | false)[]) =>
-    async (context: Context, ...args: unknown[]): Promise<unknown | ErrorResult> => {
-        let error = undefined;
+    <T extends FunctionHandler>(
+        beforeExecution: (BeforeExecutionFunction<T> | false)[],
+        handler: T,
+        postExecution: (PostExecutionFunction<T> | false)[],
+    ) =>
+    async (request: Parameters<T>, context: InvocationContext): Promise<ReturnType<T> | Error> => {
+        let handlerResult: MiddlewareResult<ReturnType<T>> = { $failed: false, $result: undefined };
 
         if (beforeExecution) {
             const middlewareFunctions = beforeExecution.filter(
-                (predicate): predicate is AzureFunction => predicate !== false,
+                (predicate): predicate is BeforeExecutionFunction<T> => predicate !== false,
             );
             for (const middlewareFunction of middlewareFunctions) {
                 try {
-                    // TODO: Give before-execution functions a parameter to indicate if the handler failed. So each function has the ability to decide for itself, if it should get executed.
-                    if (error === undefined) {
-                        await middlewareFunction(context, ...args);
-                    }
+                    await middlewareFunction(request, context, handlerResult);
                 } catch (err) {
-                    error = err;
+                    if (err instanceof Error) {
```

Comment:
Für was brauchen wir `err instanceof Error`?

Reply:
Sollten wir zur Sicherheit nicht noch eine Logik ergänzen, falls `err` kein `Error` ist?

Reply:
Hm, ich könnte einen Default (Internal-Server)-Error zurückgeben.

Reply:
Ja, da fände ich gut. Alternativ könnte man auch new Error(`Catched ${err} which is not of type Error)`) verwenden.